can_validator/MY17_Can_Library_c.py

```diff
@@ -0,0 +1,168 @@
+            "      }\n")
+        f.write(
+            "    case VCU_MC_MESSAGE__id:\n" +
+            "      if (first_byte == CAN_MC_REG_MSG_REQUEST || first_byte == CAN_MC_REG_MSG_EVENT_REQUEST) {\n" +
+            "        return Can_Vcu_MCRequest_Msg;\n" +
+            "      } else if (first_byte == CAN_MC_REG_TORQUE_CMD) {\n" +
+            "        return Can_Vcu_MCTorque_Msg;\n" +
+            "      } else {\n" +
+            "        return Can_Unknown_Msg;\n" +
+            "      }\n")
+
+        # Finish up
+        f.write(
+            "    default:\n" +
+            "      return Can_Unknown_Msg;\n" +
+            "   }\n" +
+            "}\n\n")
+
+        for message in spec.messages.values():
+            if message.name in unused_messages:
+                continue
+
+            # Write TO_CAN
+            f.write(
+                "TO_CAN(Can_" + message.name + ") {\n" +
+                "  uint64_t bitstring = 0;\n")
+
+            length = 0
+            for segment_name, segment in message.segments.items():
+                field_name = segment_name
+
+                if field_name == "unused" or field_name == "reserved":
+                    continue
+
+                if not message.is_big_endian and (segment.c_type.startswith("int") or
+                                                  segment.c_type.startswith("uint")):
+                    f.write(
+                        "  " + segment.c_type + " " + field_name + "_swap_value = swap_" + segment.c_type[:-2] +
+                        "(type_in->" + field_name + ");\n" +
+                        "  bitstring = INSERT(" + field_name + "_swap_value, bitstring, " + str(segment.position) +
+                        ", " + str(segment.length) + ");\n\n")
+                else:
+                    f.write(
+                        "  bitstring = INSERT(type_in->" + field_name + ", bitstring, " + str(segment.position) +
+                        ", " + str(segment.length) + ");\n")
+
+                length += segment.length
+            f.write(
+                "  from_bitstring(&bitstring, can_out->data);\n" +
+                "  can_out->id = " + message.name.upper() + "__id;\n" +
+                "  can_out->len = " + str(ceil(length / 8)) + ";\n" +
+                "}\n\n")
+
+            # Write FROM_CAN
+            f.write(
+                "FROM_CAN(Can_" + message.name + ") {\n" +
+                "  uint64_t bitstring = 0;\n" +
+                "  to_bitstring(can_in->data, &bitstring);\n")
+            for segment_name, segment in message.segments.items():
+                field_name = segment_name
+                if field_name == "unused" or field_name == "reserved":
+                    continue
+                if segment.c_type == "enum":
+                    # Fix name mismatch
+                    enum_name = "Can_" + message.name.replace("Heartbeat", "State") + "ID_T"
+                    if enum_name == "Can_Vcu_DashStateID_T":
```

Comment:
I know you're trying to keep backwards compatibility here, but IMO you should really just fix the problem upstream and remove all references to whatever `DashStateID` is. Having special casing in here will make debugging a huge nightmare when someone eventually tries to name something `Vcu_DashStateID` and wonders why it's not working.

Searching through github, I can't even find any references to `DashStateID` in the first place, so I don't think it should be too bad.

Reply:
Yeah, you can easily change the spec for this.

Reply:
The [BMS_Heartbeat](https://github.com/MITMotorsports/MY17_Vcu/blob/a549b2f34f68f22dc703532698819c066c3fc6cd/Input.cpp#L93), [Vcu_DashHeartbeat](https://github.com/MITMotorsports/MY17_Vcu/blob/a549b2f34f68f22dc703532698819c066c3fc6cd/Output.cpp#L271), and the [Vcu_BmsHeartbeat](https://github.com/MITMotorsports/MY17_Vcu/blob/a549b2f34f68f22dc703532698819c066c3fc6cd/Output.cpp#L320) are all referenced in VCU code (and in the dash code as well), while the state/heartbeat substitution is need for enum names (used in [vcu](https://github.com/MITMotorsports/MY17_Vcu/blob/a549b2f34f68f22dc703532698819c066c3fc6cd/Input.cpp#L223) code, for example). I can add more comments explaining this stuff, but we would have to break backwards compatibility in a significant way if I change the spec